### PR TITLE
[NVPTX] fixup support for unaligned parameters and returns

### PR DESCRIPTION
**Pull request by AlexMaclean**
Add support for unaligned parameters and return values. These must be loaded and stored one byte at a time and then bit manipulation is used to assemble the correct final result.

### CODE REVIEW

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp [hunk 1]
```diff
@@ -3045,9 +3205,28 @@ SDValue NVPTXTargetLowering::LowerFormalArguments(
...
+                    DL.getABITypeAlign(EltVT.getTypeForEVT(F->getContext()));
+              }
+            } else {
+              PartAlign =
+                  DL.getABITypeAlign(EltVT.getTypeForEVT(F->getContext()));
+            }
+            PartAlign = commonAlignment(PartAlign.valueOrOne(), Offsets[parti]);
+          } else {
+            PartAlign = MaybeAlign(aggregateIsPacked);
+          }
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:3225]: I think a lambda+early return may be a bit more readable.
```
PartsAlign = [](){
  if (NumElts != 1)
     return MaybeAlign(aggregateIsPacked);
  if (aggregateIsPacked)
     return Align(1);
  if (Offsets[parti] != 0)
     return DL.getABITypeAlign(EltVT.getTypeForEVT(F->getContext()));
  if (MaybeAlign ParamAlign = PAL.getParamAlignment(i))
      return ParamAlign.value();
    ...
  
}();
```

Comment (AlexMaclean) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:3225]: I agree this is a bit ugly. Unfortunately, I think line 3224 means that we won't be able to significantly clean it up this way. 
```c++
 PartAlign = commonAlignment(PartAlign.valueOrOne(), Offsets[parti]);
```

Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:3225]: I'm missing something here. What makes this particular line special? 

Unless I've messed something up it all boils down to this:
```
    if (NumElts != 1) return MaybeAlign(aggregateIsPacked);
    if (aggregateIsPacked) return Align(1);
   MaybeAlign PartAlign =
        (Offsets[parti] == 0 && PAL.getParamAlignment(i))
            ? PAL.getParamAlignment(i).value()
            : DL.getABITypeAlign(EltVT.getTypeForEVT(F->getContext()));
    return commonAlignment(PartAlign.valueOrOne(), Offsets[parti]);
```

Comment (AlexMaclean) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: What you have now looks right, I've updated to use that (with some small tweaks). I just meant that we wouldn't be able to do as much early exiting as you had in the first snippet.

**Comment by Artem-B** [comment]
Looks like I did commit the original test, but I suspect I may just have been a messenger and only documented the discovered issue. Thank you for taking care of it.

The patch itself LGTM.

That said, considering that we've survived for so long with nobody noticing the issue, and that the fix is nontrivial, perhaps it may be better to disallow or ignore unaligned arguments? Did you run into any real use cases where parameter/return alignment was an issue?

**Review by Artem-B — APPROVED** [review]

**Comment by AlexMaclean** [comment]
> That said, considering that we've survived for so long with nobody noticing the issue, and that the fix is nontrivial, perhaps it may be better to disallow or ignore unaligned arguments? Did you run into any real use cases where parameter/return alignment was an issue?

I think you were involved in originally reporting this issue in CUDA way back in 2017, right? Do you recall more about why it was needed at the time? It's been fixed in CUDA for a long time now and I think it would be good to add support for `__attribute__((packed))` to the open source version as well. I agree it doesn't seem like it is going to be used all that frequently but still nice to have and the fix is already complete.

**Review by Artem-B — APPROVED** [review]

**Comment by Artem-B** [comment]
> Do you recall more about why it was needed at the time?

Nope. Considering that the test landed along with the load/store vectorization patch, I suspect that I've just ran into it when I was adding other load/store tests. I do not see any internal bugs filed for that and I do not recall the issue ever coming back to my attention till now.

I'm glad to see it fixed, even if we were lucky not to be bitten by it.

**Pull request merged**